<a href="https://colab.research.google.com/github/yul77/Review/blob/main/HMLLM/%ED%9C%B4%EB%A8%BC_LLM_%EC%88%98%EC%97%85_1%EC%B0%A8_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 자연어 처리 (NLP, Natural Language Processing) introduction

- 자연어 처리(NLP)는 컴퓨터가 인간 언어를 이해하고 생성할 수 있도록 하는 기술을 연구하는 분야로, 언어학, 컴퓨터 과학, 인공지능(AI) 기술의 교차점에 위치합니다. NLP는 다양한 언어적 과업을 수행하기 위해 다양한 알고리즘, 모델 및 기법을 사용합니다. 이러한 과업은 텍스트 데이터의 구조화, 분석, 생성 등을 포함합니다.

### 주요 기술적 과업 및 기법

1. **텍스트 전처리**
    - **토큰화(Tokenization)**: 문장을 단어, 문장 또는 어절 단위로 분리합니다. 예를 들어, "NLP is fascinating."이라는 문장은 ["NLP", "is", "fascinating", "."]으로 토큰화될 수 있습니다.

In [10]:
import nltk                                                                         # nlt를 연습할 수 있게 하는 프로그램
import numpy as np

# Ensure the punkt data is downloaded
nltk.download('punkt')

# Tokenization
text = "NLP is fascinating."
tokens = nltk.word_tokenize(text)
print("Tokens:", tokens)  # Output: ['NLP', 'is', 'fascinating', '.']

# One-hot encoding representation   ( 문자를 숫자로 인코딩 원핫 인코딩 )
vocab = sorted(set(tokens))
print("Vocabulary:", vocab)  # ['.', 'NLP', 'fascinating', 'is']

one_hot_vectors = np.eye(len(vocab))[np.array([vocab.index(token) for token in tokens])]

# Display the one-hot vectors for each token
for token, vector in zip(tokens, one_hot_vectors):
    print(f"Token: {token} -> Vector: {vector}")

Tokens: ['NLP', 'is', 'fascinating', '.']
Vocabulary: ['.', 'NLP', 'fascinating', 'is']
Token: NLP -> Vector: [0. 1. 0. 0.]
Token: is -> Vector: [0. 0. 0. 1.]
Token: fascinating -> Vector: [0. 0. 1. 0.]
Token: . -> Vector: [1. 0. 0. 0.]


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


# **Q**. 오타가 나면 인식을 못한다고 하는데 그게 내가 사전에 인코딩한 문장 외의 문자를 입력하면 오류가 나는거야? 아니면 nlt 프로그램에 해당 안되는 문자를 입력했을 시 오류가 난다는거야?




 LLM 수업에서 오타로 인해 인식 오류가 나는 이유는 주로 모델이 사전에 학습한 데이터와 관련이 있습니다. 두 가지 경우로 나눠서 설명할 수 있습니다:

사전에 인코딩한 문장 외의 문자 입력: 만약 학습이나 처리할 때 사용하는 데이터나 문장이 특정한 규칙이나 포맷으로 인코딩되어 있을 때, 그 규칙을 벗어난 문자가 입력되면 인식 오류나 제대로 처리가 안 될 가능성이 큽니다. 예를 들어, 기계가 처리할 수 있는 범위 밖의 특수 문자가 들어가면 문제를 일으킬 수 있습니다.

NLT(Natural Language Toolkit) 프로그램과 관련된 문제: NLTK 같은 자연어 처리 도구는 특정 언어 패턴이나 형식에 맞춰 작동합니다. 만약 프로그램이 학습하지 않은, 또는 지원하지 않는 문자가 입력되면 예상치 못한 결과나 오류가 발생할 수 있습니다. 예를 들어, 오타나 특수문자가 입력될 경우, 그 문자를 처리할 수 없어서 잘못된 결과가 나올 수 있죠.

결론적으로, 오타나 인식 불가 문자는 모델이 학습한 내용과 실제 입력값이 불일치하거나 해당 프로그램이 지원하지 않는 문자가 포함된 경우에 발생합니다.

# **Q.그러면 nlt프로그램이 학습하지 않은 언어라는건 어떤게 있어? 한국어도 오류에 포함되는건가?**

* 정규화(Normalization): 텍스트를 일관된 형식으로 변환하는 작업입니다. 예를 들어, 모든 문자를 소문자로 변환하거나, 불필요한 기호를 제거할 수 있습니다.

In [11]:
text = "NLP is fascinating!!!"
normalized_text = text.lower().replace("!!!", "")  # 느낌표를 블랭크로 대체하라는 명령어. = 느낌표를 제거하라 = 원핫인코딩을 위해 느낌표 마침표등을 제거한다.
print(normalized_text)  # Output: 'nlp is fascinating' -> text.lower() 사용하려 소문자로만 출력된다

nlp is fascinating


* 형태소 분석(Morphological Analysis): 단어의 어근, 접두사, 접미사를 분석하여 단어의 의미를 파악합니다. 예를 들어, "running"이라는 단어는 "run"이라는 어근과 "-ing" 접미사로 분해됩니다.

In [12]:
from nltk.stem import PorterStemmer
stemmer = PorterStemmer()
word = "running"
stem = stemmer.stem(word)
print(stem)  # Output: 'run'  = ing가 들어간 다른 단어들을 넣어도 원형으로 출력이 된다 ex)  flying , moving 등등
             # nltk 어휘사전에 등록되어있어서 원형으로 나온다


run


* 품사 태깅(Part-of-Speech Tagging): 각 단어에 대해 문법적 품사를 지정하는 작업입니다. 예를 들어, "He is running."이라는 문장에서 "He"는 대명사(PRP), "is"는 동사(VBZ), "running"은 동명사(VBG)로 태깅됩니다.

In [13]:
import nltk
nltk.download('averaged_perceptron_tagger')
sentence = "He is running."
pos_tags = nltk.pos_tag(word_tokenize(sentence))
print(pos_tags)  # Output: [('He', 'PRP'), ('is', 'VBZ'), ('running', 'VBG')]

# 오류 맞음 - nltk에서 'word_tokenize'를 더이상 안쓰는 것 같다고 새로운 코드로 알려주시겠다고 함.

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


NameError: name 'word_tokenize' is not defined

In [17]:
# gpt한테 물어봐서 수정한 코드
import nltk
nltk.download('averaged_perceptron_tagger')
from nltk.tokenize import word_tokenize # import the word_tokenize function
sentence = "He is running."
pos_tags = nltk.pos_tag(word_tokenize(sentence))
print(pos_tags)  # Output: [('He', 'PRP'), ('is', 'VBZ'), ('running', 'VBG')]

[('He', 'PRP'), ('is', 'VBZ'), ('running', 'VBG'), ('.', '.')]


[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


2. **언어 모델링(Language Modeling)**
    - **n-그램 모델**: n-그램 모델은 단어 시퀀스의 확률을 계산하는 모델입니다. 예를 들어, 2-그램(bi-gram) 모델은 두 단어 쌍의 확률을 계산합니다. "I love NLP"에서 2-그램 모델은 P("love" | "I")와 P("NLP" | "love")의 확률을 계산하여 전체 문장의 확률을 평가합니다.
    - **신경망 언어 모델**:
        - **RNN (Recurrent Neural Networks)**: 시퀀스 데이터를 처리하는 신경망으로, 이전 상태를 고려하여 다음 단어를 예측합니다. 그러나 RNN은 긴 문맥을 처리하는 데 한계가 있습니다.
        - **LSTM (Long Short-Term Memory)**: RNN의 변형으로, 장기 의존성을 처리할 수 있도록 설계되었습니다. LSTM은 "셀 상태"를 유지하며, 이전 정보가 더 오래 유지될 수 있도록 합니다.
        - **Transformer**: 자기 주의 메커니즘(Self-Attention)을 사용하는 모델로, 병렬 처리가 가능하며, 긴 문맥을 효과적으로 처리할 수 있습니다. BERT(Bidirectional Encoder Representations from Transformers)와 GPT(Generative Pre-trained Transformer)는 Transformer 아키텍처에 기반한 대표적인 모델입니다.

<span style="color:blue"> Ai 는 룰베이스를 기반으로 한다. 인공지능은 스스로 판단할 수 있어야 하고 if else까지 포함된다.

룰베이스의 한계를 극복하고자 개발된것 = 머신러닝
머신러닝 : 딥러닝,수학적알고리즘 (SVN,DT,KNN)
우리가 공부할 것은 딥러닝에 속한다.


**RNN** = 시간을 메모리에 입력한것. 단 메모리가 짧은 것이 단점. 과거에 있었던 말을 기억을 못하는 문제 발생. 한문장은 괜찮은데 문단은 처리 불가.

그래서 만들어진 알고리즘이 **LSTM**

</span>



In [15]:
from transformers import GPT2Tokenizer, GPT2LMHeadModel
import torch

tokenizer = GPT2Tokenizer.from_pretrained("gpt2")  # gpt2 모델로 토크나이즈 한다.
model = GPT2LMHeadModel.from_pretrained("gpt2")

input_text = "Natural language processing is"
input_ids = tokenizer.encode(input_text, return_tensors="pt")
output = model.generate(input_ids, max_length=50, num_return_sequences=1)

generated_text = tokenizer.decode(output[0], skip_special_tokens=True)
print(generated_text)

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


Natural language processing is a very important part of the language learning process.

The first step is to understand the language. The second step is to understand the language.

The first step is to understand the language.

The second step


GPT모델 = 생성을 하겠다

특징 ->피쳐

3. **정보 검색(Information Retrieval) 및 추출(Extraction)**
    - **TF-IDF (Term Frequency-Inverse Document Frequency)**: 문서 내 특정 단어의 중요성을 평가하는 기법입니다. 단어의 빈도(Term Frequency)와 해당 단어가 등장한 문서 수의 역수(Inverse Document Frequency)를 곱하여 단어의 가중치를 계산합니다.
        - **TF-IDF**는 Python의 `sklearn` 라이브러리를 사용하여 TF-IDF를 계산합니다.
            - 개념
                - **TF (Term Frequency)**: 특정 단어가 문서 내에서 얼마나 자주 등장하는지를 나타내는 값.
                - **IDF (Inverse Document Frequency)**: 특정 단어가 전체 문서 집합에서 얼마나 흔하지 않은지를 나타내는 값.
            
            TF-IDF는 단어의 중요도를 평가하여, 문서에서 자주 등장하지만 모든 문서에 공통적으로 나타나는 단어들(예: "the", "and")의 중요도를 낮추고, 특정 문서에서만 빈번하게 등장하는 단어들의 중요도를 높입니다.

In [16]:
from sklearn.feature_extraction.text import TfidfVectorizer

# 예제 문서 집합
documents = [
    "자연어 처리는 인공지능의 한 분야입니다.",
    "자연어 처리와 컴퓨터 비전은 인공지능의 두 가지 중요한 분야입니다.",
    "컴퓨터 비전은 이미지와 비디오 데이터를 처리하는 기술입니다."
]

# TF-IDF 벡터라이저 초기화
vectorizer = TfidfVectorizer()                                                   # 토큰화

# 문서에 TF-IDF 적용
tfidf_matrix = vectorizer.fit_transform(documents)                               #특정 단어의 빈도수 분석

# 단어 리스트 출력
print("Vocabulary:", vectorizer.get_feature_names_out())

# TF-IDF 행렬 출력
print("TF-IDF Matrix:\n", tfidf_matrix.toarray())

# 특정 단어의 TF-IDF 값 출력 (예: '인공지능의')
word = '인공지능의'
word_index = vectorizer.vocabulary_.get(word)
if word_index is not None:
    print(f"\nTF-IDF value for '{word}':", tfidf_matrix[:, word_index].toarray())
else:
    print(f"'{word}' is not in the vocabulary.")

Vocabulary: ['가지' '기술입니다' '데이터를' '분야입니다' '비디오' '비전은' '이미지와' '인공지능의' '자연어' '중요한' '처리는'
 '처리와' '처리하는' '컴퓨터']
TF-IDF Matrix:
 [[0.         0.         0.         0.45985353 0.         0.
  0.         0.45985353 0.45985353 0.         0.60465213 0.
  0.         0.        ]
 [0.41197298 0.         0.         0.31331607 0.         0.31331607
  0.         0.31331607 0.31331607 0.41197298 0.         0.41197298
  0.         0.31331607]
 [0.         0.40301621 0.40301621 0.         0.40301621 0.30650422
  0.40301621 0.         0.         0.         0.         0.
  0.40301621 0.30650422]]

TF-IDF value for '인공지능의': [[0.45985353]
 [0.31331607]
 [0.        ]]


*  띄어쓰기 기준으로 잘라준다 .
*  특징값 [[0.45985353] 이라고 나온다.
* TF-IDF로도 충분하지만 좀 더 고도화 된 작업을 할 때 BM25 사용하는 것도 괜찮다.


- **TF-IDF Matrix**: 각 단어가 각 문서에서 가지는 중요도를 나타냅니다. 값이 클수록 해당 문서에서 더 중요한 단어로 간주됩니다.
- 예를 들어, `인공지능`이라는 단어의 TF-IDF 값을 확인하면, 특정 문서에서 얼마나 중요한지 알 수 있습니다.

- **BM25**: TF-IDF의 확장 버전으로, 단어의 빈도와 문서 길이를 고려하여 검색의 정확도를 높이는 정보 검색 모델입니다. **BM25는** `rank-bm25` 라이브러리를 사용하여 BM25를 계산 합니다.
    
    개념
    
    - **BM25**: TF-IDF의 개선된 버전으로, 단어의 빈도, 문서 길이, 그리고 TF와 IDF 간의 균형을 고려하여 검색 성능을 최적화한 정보 검색 알고리즘입니다. BM25는 단순 TF-IDF보다 더 정교한 방법으로 문서와 질의 간의 유사도를 측정합니다.

In [18]:
! pip install rank-bm25

from rank_bm25 import BM25Okapi
from nltk.tokenize import word_tokenize

# 예제 문서 집합
documents = [
    "자연어 처리는 인공지능의 한 분야입니다.",
    "자연어 처리와 컴퓨터 비전은 인공지능의 두 가지 중요한 분야입니다.",
    "컴퓨터 비전은 이미지와 비디오 데이터를 처리하는 기술입니다."
]

# 문서를 토큰화
tokenized_documents = [word_tokenize(doc) for doc in documents]

# BM25 모델 초기화
bm25 = BM25Okapi(tokenized_documents)

# 예제 질의
query = "인공지능 분야"

# 질의를 토큰화
tokenized_query = word_tokenize(query)

# 문서들의 BM25 점수 계산
scores = bm25.get_scores(tokenized_query)

# 각 문서에 대한 BM25 점수 출력
print("BM25 Scores:", scores)

# 가장 관련성 높은 문서 출력
best_doc_index = scores.argmax()
print("\\nMost relevant document:", documents[best_doc_index])  # TF-IDF 보다 괜찮은 결과가 나온다. 협업에 가서도 이 코드 그대로 사용 가능
                                                                # 다른 문장들을 입력해서 여러가지 시도도 해볼 것

BM25 Scores: [0. 0. 0.]
\nMost relevant document: 자연어 처리는 인공지능의 한 분야입니다.


In [22]:
# 연습

! pip install rank-bm25

from rank_bm25 import BM25Okapi
from nltk.tokenize import word_tokenize

# 예제 문서 집합
documents = [
    "자연어 처리는 인공지능의 한 분야입니다.",
    "자연어 처리와 컴퓨터 비전은 인공지능의 두 가지 중요한 분야입니다.",
    "컴퓨터 비전은 이미지와 비디오 데이터를 처리하는 기술입니다."
]

# 문서를 토큰화
tokenized_documents = [word_tokenize(doc) for doc in documents]

# BM25 모델 초기화
bm25 = BM25Okapi(tokenized_documents)

# 예제 질의
query = "컴퓨터"

# 질의를 토큰화
tokenized_query = word_tokenize(query)

# 문서들의 BM25 점수 계산
scores = bm25.get_scores(tokenized_query)

# 각 문서에 대한 BM25 점수 출력
print("BM25 Scores:", scores)

# 가장 관련성 높은 문서 출력
best_doc_index = scores.argmax()
print("\\nMost relevant document:", documents[best_doc_index])

BM25 Scores: [0.         0.01479238 0.01645652]
\nMost relevant document: 컴퓨터 비전은 이미지와 비디오 데이터를 처리하는 기술입니다.


- **BM25 Scores**: 각 문서가 주어진 질의와 얼마나 관련성이 있는지를 나타냅니다. 값이 클수록 해당 문서가 질의와 더 관련이 있다고 판단됩니다.
- 예제에서는 "인공지능 분야"라는 질의를 기준으로 문서들의 관련성 점수를 계산하여, 가장 관련성이 높은 문서를 출력합니다.
- **Named Entity Recognition (NER)**: 텍스트에서 사람, 장소, 조직 등 명명된 실체를 추출하는 작업입니다.

In [ ]:
from transformers import pipeline

nlp = pipeline("ner")
text = "John works at OpenAI."
entities = nlp(text)
print(entities)  # Output: [{'word': 'John', 'score': 0.99, 'entity': 'PER'}, ...]

4. **자연어 생성(NLG, Natural Language Generation)**
    - **Seq2Seq 모델**: 입력 시퀀스에서 출력 시퀀스로 변환하는 모델로, 기계 번역에서 자주 사용됩니다. 입력 문장을 받아 번역된 문장을 생성하거나 요약을 생성할 수 있습니다.
    - **Transformer 기반 생성**: GPT 시리즈와 같은 모델은 문장을 생성하는 데 사용되며, 텍스트 완성, 스토리 생성 등 다양한 응용 분야에 사용됩니다.
5. **기계 번역(Machine Translation)**
    - **통계적 기계 번역(Statistical Machine Translation, SMT)**: 문장 단위로 번역 확률을 계산하는 기법입니다. 입력 언어와 출력 언어 간의 번역 패턴을 통계적으로 학습합니다.
    - **신경망 기계 번역(Neural Machine Translation, NMT)**: Seq2Seq 모델과 Attention 메커니즘을 사용하여 문장을 번역합니다. 현재는 Transformer 기반의 NMT가 널리 사용됩니다.

In [ ]:
from transformers import MarianMTModel, MarianTokenizer

model_name = "Helsinki-NLP/opus-mt-en-de"
tokenizer = MarianTokenizer.from_pretrained(model_name)
model = MarianMTModel.from_pretrained(model_name)

input_text = "Natural language processing is fascinating."
translated = model.generate(**tokenizer(input_text, return_tensors="pt", padding=True))
translation = tokenizer.decode(translated[0], skip_special_tokens=True)
print(translation)  # Output: 'Die Verarbeitung natürlicher Sprache ist faszinierend.'